Hive: Add support for custom catalog to Iceberg StorageHandler (#1155)

guilload opened the pull request:
This PR adds support for custom catalog to Iceberg StorageHandler as well as Hadoop and Hive catalogs and removes the duplicated logic in `mapreduce.IcebergInputFormat.findTable` and `mapred.TableResolver`.

Ideally, I'd like to run the Hive test suite against all the catalog implementations, but I didn't find a way to combine the `HiveRunner` and `Parameterized` test runners together. Suggestions?

@cmathiesen @massdosage @rdsr

guilload commented on mr/src/main/java/org/apache/iceberg/mr/InputFormatConfig.java:
```diff
@@ -95,15 +93,12 @@ public ConfigBuilder schema(Schema schema) {
+      return this;
     }
 
-    public ConfigBuilder readFrom(String path) {
```
I have decided to remove `readFrom(String ...)` since the string can be ambiguously an identifier or a location.

rdblue commented:
@guilload, you can use an abstract class for the test cases and then multiple concrete classes for the environments you want to run with.

guilload commented:
Yeah, if we don't mind having as many test classes as catalog implementations then I can do that. 

Also, is the running time of the Iceberg test suite something we care about? `HiveRunner` is rather slow (I believe each test sets up a new local HMS from scratch) and `TestHiveIcebergInputFormat` takes about 4 minutes to run on my machine (compile time included).  Running that test with more catalog implementations is only going to increase that.

rdblue commented:
It's not ideal to have slow tests, but better to have slow tests than leave things untested, I think.

rdsr commented on mr/src/main/java/org/apache/iceberg/mr/CatalogLoader.java:
```diff
@@ -0,0 +1,29 @@
+public interface CatalogLoader {
+
+  Catalog load(Configuration conf);
+
```
nit: blank line. Also javadoc on this method

rdsr commented on mr/src/main/java/org/apache/iceberg/mr/Catalogs.java:
```diff
@@ -0,0 +1,110 @@
+    // A bit of a hack to make this function work transparently with Hive without having to remap the "name" and
+    // "location" properties.
+    if (HiveConf.getVar(conf, HiveConf.ConfVars.HIVEQUERYID).length() > 0) {
+      return loadTable(conf, conf.get(NAME), conf.get(LOCATION));
```
How's the name property used to specify a table ?

pvary replied:
Where is this conf coming from? Is this a query conf? Or this is some feature of Hive I am not aware of? Then this does not seem too future proof. What happens if there are multiple Iceberg tables are in a single query?

massdosage replied:
A useful test could be to add a HiveRunner test which does a join between multiple tables, each one loaded with a different catalog. We definitely found some issues in our earlier implementation of this that didn't hold once more than one table was involved in the query.

rdsr commented on mr/src/main/java/org/apache/iceberg/mr/InputFormatConfig.java:
```diff
@@ -95,15 +93,12 @@ public ConfigBuilder schema(Schema schema) {
 
-    public ConfigBuilder readFrom(String path) {
-      conf.set(TABLE_PATH, path);
+    public ConfigBuilder readFrom(File location) {
```
Shouldn't this be `InputFile`

rdblue replied:
I don't think so. I think it should be `String location` because it identifies a table by location (like HadoopTables) right?

rdsr replied:
Yes, you are right! I just wanted this to not be a `java.io.File` type

pvary commented:
> `HiveRunner` is rather slow (I believe each test sets up a new local HMS from scratch) and `TestHiveIcebergInputFormat` takes about 4 minutes to run on my machine (compile time included).

In most of our test cases we spin up HMS once per test file and clean it up in setUp/tearDown. Not an ideal solution, but saves as time/money.

massdosage commented:
> This PR adds support for custom catalog to Iceberg StorageHandler as well as Hadoop and Hive catalogs and removes the duplicated logic in `mapreduce.IcebergInputFormat.findTable` and `mapred.TableResolver`.
> 
> Ideally, I'd like to run the Hive test suite against all the catalog implementations, but I didn't find a way to combine the `HiveRunner` and `Parameterized` test runners together. Suggestions?
> 
> @cmathiesen @massdosage @rdsr

I've seen this doing using JUnit 5 - it would look something like this so it should be possible:

```
@ExtendWith(HiveRunnerExtension.class)
public class Params {
  
  @HiveSQL(files = {})
  private HiveShell shell;

  @ParameterizedTest
  @ValueSource(strings = { "orc", "parquet", "avro" })
  public void fileTypes(String fileType) {
    System.out.println(fileType);
  }
  
}
```
It's probably not possible with JUnit4 as both mechanisms use a Runner and I don't think you can have multiple Runners on one test class.

massdosage commented:
> > `HiveRunner` is rather slow (I believe each test sets up a new local HMS from scratch) and `TestHiveIcebergInputFormat` takes about 4 minutes to run on my machine (compile time included).
> 
> In most of our test cases we spin up HMS once per test file and clean it up in setUp/tearDown. Not an ideal solution, but saves as time/money.

HiveRunner doesn't currently support that but it has been discussed before - https://github.com/klarna/HiveRunner/issues/69

guilload commented:
Hey,

I've just pushed a commit that adds the changes requested during the first round of review and additionally integrates the modifications suggested by @rdsr in #1267, which clean up the input format nicely and solve the issue reported in #1267.

I should be able to add the tests for each catalog today or tomorrow.

cc @massdosage

rdsr commented on mr/src/main/java/org/apache/iceberg/mr/Catalogs.java:
```diff
@@ -0,0 +1,103 @@
+   * @return an Iceberg table
+   */
+  public static Table loadTable(Configuration conf) {
+    return loadTable(conf, conf.get(InputFormatConfig.TABLE_IDENTIFIER), conf.get(InputFormatConfig.TABLE_LOCATION));
```
can't we replace TABLE_LOCATION, TABLE_IDENTIFIER and TABLE_PATH with just TABLE_PATH?

e.g see org.apache.iceberg.spark.source.IcebergSource#findTable

rdsr replied:
We can ignore this in light of further comments below.

rdsr commented on mr/src/main/java/org/apache/iceberg/mr/Catalogs.java:
```diff
@@ -0,0 +1,103 @@
+    return loadTable(conf, props.getProperty(NAME), props.getProperty(LOCATION));
+  }
+
+  private static Table loadTable(Configuration conf, String tableIdentifier, String tableLocation) {
```
Does it make sense to do something similar to what spark does?
```
protected Table findTable(Map<String, String> options, Configuration conf) {
    Preconditions.checkArgument(options.containsKey("path"), "Cannot open table: path is not set");
    String path = options.get("path");

    if (path.contains("/")) {
      HadoopTables tables = new HadoopTables(conf);
      return tables.load(path);
    } else {
      HiveCatalog hiveCatalog = HiveCatalogs.loadCatalog(conf);
      TableIdentifier tableIdentifier = TableIdentifier.parse(path);
      return hiveCatalog.loadTable(tableIdentifier);
    }
  }
```
use TABLE_PATH to determine whether we need to use Hadoop or Hive catalog first and only then try custom catalog in the end. Seems like custom catalog would be used very rarely, if at **all.**

rdblue replied:
I think what's happening here makes sense because configuration is done separately using either `readFrom(location)` or `readFrom(identifier)`.

rdsr commented on mr/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergSerDe.java:
```diff
@@ -39,13 +38,7 @@
-    } catch (IOException e) {
-      throw new UncheckedIOException("Unable to resolve table from configuration: ", e);
-    }
+    Table table = Catalogs.loadTable(configuration, serDeProperties);
```
might require a rebase

rdblue commented on mr/src/main/java/org/apache/iceberg/mr/Catalogs.java:
```diff
@@ -0,0 +1,103 @@
+    if (catalogName != null) {
+      switch (catalogName.toLowerCase()) {
+        case HADOOP:
+          return Optional.of(new HadoopCatalog(conf));
```
There are two relevant configs in addition to the type: `uri` for Hive and `warehouse` for Hadoop. Hive defaults to the value in configuration of `hive.metastore.uris` from `hive-site.xml`, so it makes sense to not pass it explicitly here.

But, the Hadoop catalog doesn't really have a standard warehouse path. It just has a default. Should we add a configuration option to set up the warehouse when using a `HadoopCatalog`?

guilload replied:
Yes, that would be useful. I've just pushed 90ccd39b8e349b3506c9866e179112a408447bfa for that purpose.

rdblue commented on mr/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java:
```diff
@@ -66,7 +73,12 @@ public HiveAuthorizationProvider getAuthorizationProvider() throws HiveException
 
+    map.put(InputFormatConfig.TABLE_IDENTIFIER, props.getProperty(NAME));
+    map.put(InputFormatConfig.TABLE_LOCATION, table.location());
+    map.put(InputFormatConfig.TABLE_SCHEMA, SchemaParser.toJson(table.schema()));
```
Doesn't this also need to forward the catalog loader class?

rdsr replied:
Depends upon where we set the catalog loader class.   If we set it in HiveConf, then it will always be available by default in configuration. If we want to expose catalog loader class as a table property, e.g to load different tables differently, then we can forward it here.

rdblue replied:
It seems like it would always comes from table properties to me, since we might have tables from different Iceberg catalogs.

massdosage replied:
Yes, and ideally we'd have a unit test which uses multiple tables which are set up differently and checks that the properties that are unique to each table are honoured properly.

guilload replied:
The current implement of the catalog loader reads directly from the HiveConf and ignores the table properties, that why the catalog loader class is not forwarded here.

I'm planning to add the ability to declare the catalog in the table properties but I'd rather do so in a follow-up PR because that requires a refactor of the current test suite. I'll make sure to add a test that reads from multiples tables set up with different catalogs.

rdblue commented:
I like the approach, it mostly looks good to me. Would be great to have @massdosage check that the behavior is what they expect, too.

massdosage commented:
I've managed to run this branch successfully on Hive in distributed mode (just with some packaging changes to build an uber jar and do some relocations) and it works fine. I'll keep checking when there are major changes but so far it looks good.

pvary commented:
> > > `HiveRunner` is rather slow (I believe each test sets up a new local HMS from scratch) and `TestHiveIcebergInputFormat` takes about 4 minutes to run on my machine (compile time included).
> > 
> > 
> > In most of our test cases we spin up HMS once per test file and clean it up in setUp/tearDown. Not an ideal solution, but saves as time/money.
> 
> HiveRunner doesn't currently support that but it has been discussed before - [klarna/HiveRunner#69](https://github.com/klarna/HiveRunner/issues/69)

Hive does this here:
https://github.com/apache/hive/blob/6267520a50ba8c12ae5cee8fb27a6746376a21c6/itests/util/src/main/java/org/apache/hadoop/hive/ql/QTestUtil.java#L444

I think the only part we are interested in here is:
https://github.com/apache/hive/blob/6267520a50ba8c12ae5cee8fb27a6746376a21c6/itests/util/src/main/java/org/apache/hadoop/hive/ql/QTestUtil.java#L336

Which is very similar than what was done in klarna/HiveRunner#69. We manually went through and deleted the tables in the databases (I think there was some issue previously which should be fixed now). The only other difference is that we clean up the files as well which is needed for EXTERNAL tables. Since the Iceberg tables will be EXTERNAL, I think we will need that as well here.

massdosage commented:
> > > > `HiveRunner` is rather slow (I believe each test sets up a new local HMS from scratch) and `TestHiveIcebergInputFormat` takes about 4 minutes to run on my machine (compile time included).
> > > 
> > > 
> > > In most of our test cases we spin up HMS once per test file and clean it up in setUp/tearDown. Not an ideal solution, but saves as time/money.
> > 
> > 
> > HiveRunner doesn't currently support that but it has been discussed before - [klarna/HiveRunner#69](https://github.com/klarna/HiveRunner/issues/69)
> 
> Hive does this here:
> https://github.com/apache/hive/blob/6267520a50ba8c12ae5cee8fb27a6746376a21c6/itests/util/src/main/java/org/apache/hadoop/hive/ql/QTestUtil.java#L444
> 
> I think the only part we are interested in here is:
> https://github.com/apache/hive/blob/6267520a50ba8c12ae5cee8fb27a6746376a21c6/itests/util/src/main/java/org/apache/hadoop/hive/ql/QTestUtil.java#L336
> 
> Which is very similar than what was done in [klarna/HiveRunner#69](https://github.com/klarna/HiveRunner/issues/69). We manually went through and deleted the tables in the databases (I think there was some issue previously which should be fixed now). The only other difference is that we clean up the files as well which is needed for EXTERNAL tables. Since the Iceberg tables will be EXTERNAL, I think we will need that as well here.

That sounds promising. I guess the best place to do this is in HiveRunner itself. That issue referenced above has a PR against it from one of the original committers on HiveRunner but he makes it clear that it's very experimental and it hasn't been worked on in a long time. We could either build on that or raise a new one. I'd be happy to work with you on it if you want?

rdblue commented:
@guilload, could you rebase and take a look at the [review comment about multiple catalogs](https://github.com/apache/iceberg/pull/1243#discussion_r463838000)?

guilload commented:
@rdblue I've rebased the PR, added a commit for customizing the warehouse location for the Hadoop catalog and reviewed the comment about multiples catalogs. However, I'm still working on running the Hive test suite once per catalog implementation. I've made good progress but still need more time.

guilload commented:
I've just added 2500efb, which run the Hive test suite against multiple catalog implementations.

rdsr commented:
thanks @guilload @rdblue I plan to spend time on this today

rdsr commented on mr/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergInputFormat.java:
```diff
@@ -45,9 +47,9 @@
 
 @RunWith(StandaloneHiveRunner.class)
-public class TestHiveIcebergInputFormat {
+public abstract class TestHiveIcebergInputFormat {
```
it is kinda weird that we are calling this class `TestHiveIcebergInputFormat` instead of say `TestHiveStorageHandler` or something similar.

guilload replied:
That's because this test class was added before the storage handler. I'll rename it.

rdsr commented on mr/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergInputFormat.java:
```diff
@@ -155,11 +157,31 @@ public void testJoinTables() {
+  }
+
+  protected void createHiveTable(String tableName, String location) {
     shell.execute(String.format(
```
shouldn't we set the  custom catalog loader  by setting a table property `catalog.loader.class` here?

rdsr replied:
I think if we make `createHiveTable` [and maybe `createIcebergTable`] abstract, each implementation can set the right configuration in table properties. We can then test custom catalog and other configs without explicitly passing as properties from `TestTables` . What do you think?

rdsr replied:
If that works [maybe we also have to abstract `Configuration creation`] then I think we can drop the `TestTables` class as well

guilload replied:
Yeah, I think that will work once the catalog loader handles catalog settings defined at the table properties level. I'll add this refactor in the follow up PR.

rdsr commented:
generally LG! I'm ok apart from some comments on the tests. We can even address these comments on the tests later if we like.

rdsr commented:
Thanks @guilload lets improve the test in follow ups.
